packages/langium/src/default-module.ts

```diff
@@ -30,35 +30,58 @@ import { DefaultNameProvider } from './references/naming';
 import { DefaultReferences } from './references/references';
 import { DefaultScopeComputation, DefaultScopeProvider } from './references/scope';
 import { DefaultJsonSerializer } from './serializer/json-serializer';
-import { LangiumDefaultServices, LangiumServices } from './services';
+import { DeepPartial, LangiumDefaultServices, LangiumDefaultSharedServices, LangiumGeneratedServices, LangiumGeneratedSharedServices, LangiumServices, LangiumSharedServices } from './services';
 import { DefaultDocumentValidator } from './validation/document-validator';
 import { ValidationRegistry } from './validation/validation-registry';
+import { ExtensionServiceRegistry, ServiceRegistry, SingleServiceRegistry } from './service-registry';
 
-export type DefaultModuleContext = {
+export interface DefaultModuleContext {
     connection?: Connection
+    serviceRegistry?: ServiceRegistry
 }
 
-export function createDefaultModule(context: DefaultModuleContext = {}): Module<LangiumServices, LangiumDefaultServices> {
+export function injectSingleService<T extends LangiumServices>(
+    sharedModule: Module<LangiumSharedServices, LangiumDefaultSharedServices>,
+    generatedSharedModule: Module<LangiumSharedServices, LangiumGeneratedSharedServices>,
+    generatedModule: Module<LangiumServices, LangiumGeneratedServices>,
+    module?: Module<T, DeepPartial<T>>): T {
+
+    const shared = inject(sharedModule, generatedSharedModule);
+    const single = inject(createDefaultModule(shared), generatedModule, module);
+    shared.ServiceRegistry = new SingleServiceRegistry(single);
+    return single;
+}
+
+export function createSharedModule(context: DefaultModuleContext = {}): Module<LangiumSharedServices, LangiumDefaultSharedServices> {
     return {
-        parser: {
-            GrammarConfig: (injector) => createGrammarConfig(injector),
-            LangiumParser: (injector) => createLangiumParser(injector),
-            ValueConverter: () => new DefaultValueConverter(),
-            TokenBuilder: () => new DefaultTokenBuilder()
+        ServiceRegistry: () => context.serviceRegistry ?? new ExtensionServiceRegistry(),
```

Comment:
What's the use case for creating the registry externally?

Reply:
Simple example might be something like special build files, which are not only identified by their file extension. See `package.json` which gets special handling by vscode (code completion, validation, etc). Or even files without extensions (`BUILD` files).

Reply:
Well, we have some sort of circular dependency. An instanceof `LangiumServices` needs a shared services instance, but the registry needs the `LangiumServices` instance. By providing the registry externally, we can provide the created `LangiumServices` instance back to the service registry.